ctk/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt

```diff
@@ -15,160 +15,105 @@ package org.codice.compliance.verification.core
 import org.codice.compliance.SAMLCore_3_2_2_d
 import org.codice.compliance.SAMLCore_3_2_2_e
-import org.codice.compliance.SAMLCore_3_4
 import org.codice.compliance.recursiveChildren
 import org.codice.compliance.children
 import org.codice.compliance.utils.TestCommon.Companion.SAML_VERSION
 import org.codice.compliance.utils.TestCommon.Companion.TOP_LEVEL_STATUS_CODES
 import org.w3c.dom.Node
 
-class ResponseProtocolVerifier(private val response: Node,
-                               private val id: String,
-                               private val acsUrl: String?) {
+abstract class ResponseProtocolVerifier(open val response: Node,
+                                        open val id: String,
+                                        open val acsUrl: String?) {
     companion object {
-        private const val ID = "ID"
-        private const val VALUE = "Value"
-        private const val STATUS = "Status"
         private const val VERSION = "Version"
-        private const val RESPONSE = "Response"
+        private const val DESTINATION = "Destination"
         private const val STATUS_CODE = "StatusCode"
-        private const val ISSUE_INSTANT = "IssueInstant"
-        private const val SAMLCore_3_2_2 = "SAMLCore.3.2.2"
     }
 
-    /**
-     * Verify protocols against the Core Spec document
-     * 3.2.2 Complex Type StatusResponseType
-     */
-    fun verify() {
+    abstract fun verify()
+
+    /** 3.2.2 Complex Type StatusResponseType */
+    fun verifyCoreResponse() {
         CoreVerifier(response).verify()
         verifyStatusResponseType()
-        verifyStatusesType()
-        verifyNameIdMappingResponse()
-
-        if (response.localName == "AuthnRequest") {
-            response.children("Assertion")
-                    .forEach {
-                        if (it.children("AuthnStatement").isEmpty())
-                            throw SAMLComplianceException.create(SAMLCore_3_4,
-                                    message = "AuthnStatement not found.",
-                                    node = response)
-                    }
-        }
+        verifyStatusType()
+        verifyStatusMessage()
+        verify()
     }
 
-    /**
-     * Verify the Status Response Type
-     * 3.2.2 Complex Type StatusResponseType
-     *
-     * All SAML responses are of types that are derived from the StatusResponseType complex type.
-     */
+    /** All SAML responses are of types that are derived from the StatusResponseType complex type.*/
     private fun verifyStatusResponseType() {
-        if (response.attributes.getNamedItem(ID) == null)
-            throw SAMLComplianceException.createWithXmlPropertyReqMessage(SAMLCore_3_2_2,
-                    property = ID,
-                    parent = RESPONSE,
-                    node = response)
-        CommonDataTypeVerifier.verifyIdValues(response.attributes.getNamedItem(ID),
+        CommonDataTypeVerifier.verifyIdValues(response.attributes.getNamedItem("ID"),
```

Comment:
✏️ `response.attributes` is accessed here without safe nav. So either it is safe (and the safe accesses on lines 56, 76, and 91 don't need to be safe) or it will throw an NPE and it and the accesses on lines 64 and 74 need to be safe as well.